### PR TITLE
[ci] Add MacOS & Python sanity check for release automation pipeline

### PR DESCRIPTION
**Pull request by khluu**
## Changes
- Add `macos` and `macos-arm64` runner queues
- Add script to install `miniconda3` and run Python sanity check for MacOS.
- Add steps for MacOS `x86_64` and `arm64`. 
- MacOS Buildkite steps run on [our MacOS sandbox](https://github.com/ray-project/macos-sandbox-buildkite-plugin) with the plugin `ray-project/macos-sandbox-#v1.0.7` configured in the convert step in `ray-project/rayci`. When the steps in `wheels.rayci.yml` get converted to Buildkite steps in `rayci`, they are added this plugin.

## Checks
https://buildkite.com/ray-project/release-automation/builds/48 passes for both Mac `x86_64` and `arm64` on Python sanity check.

### CODE REVIEW

**Comment by khluu** [comment]
Moving this PR to draft for now since I'm trying out another way to install dependencies (`install-dependencies.sh`)

File: .buildkite/release-automation/verify-macos-wheels.sh [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+pip install \
+    --index-url https://test.pypi.org/simple/ \
+    --extra-index-url https://pypi.org/simple \
+    "ray[cpp]==$RAY_VERSION"
+
+(
+    cd release/util
+    python sanity_check.py
+)
```
[the remark this answers is not in the document]
Comment (khluu) [reply .buildkite/release-automation/verify-macos-wheels.sh:45]: Ah I see. That's why I couldn't find a stack but only an auto-scaling group for the Mac machines used to run on `postmerge` on AWS.
Will fix this!

[the remark this answers is not in the document]
Comment (khluu) [reply .buildkite/release-automation/verify-macos-wheels.sh:45]: @aslonnie  Besides dependencies installation (bazel, miniconda, etc.), should I put anything else in the tmp directory? I think ray repo is already there on the machines.

**Comment by khluu** [comment]
@aslonnie  I have:
- Created a tmp dir and install `miniconda` in it.
- Add a function to remove this tmp dir. This function is triggered upon exit (when the script finishes or when there's an error) which should delete all relevant files from `miniconda`.
- Run all python versions in the same script, and each step on Buildkite should only call the script once now.
- https://buildkite.com/ray-project/release-automation/builds/48 passes for all versions on Python sanity check. 
- Bazel installation & C++ sanity check is in this follow up PR #43131

File: .buildkite/release-automation/verify-macos-wheels.sh [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+TMP_DIR=$(mktemp -d "$HOME/tmp.XXXXXXXXXX")
+
+install_miniconda
+
+# Install Ray & run sanity checks for each python version
+for python_version in "${python_versions[@]}"; do
+    run_sanity_check "$python_version"
+done
+
+trap _clean_up EXIT
```
Comment (aslonnie) [inline .buildkite/release-automation/verify-macos-wheels.sh:64]: you want to move the `trap` up, ideally right after `mktemp`

File: .buildkite/release-automation/verify-macos-wheels.sh [hunk 3]
```diff
@@ -0,0 +1,64 @@
+#!/bin/bash
+
+set -euo pipefail
+
+set -x
+
+python_versions=("3.8" "3.9" "3.10" "3.11")
```
Comment (aslonnie) [inline .buildkite/release-automation/verify-macos-wheels.sh:7]: nit: caps for bash vars: `PYTHON_VERSIONS`

File: .buildkite/release-automation/verify-macos-wheels.sh [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    # Add miniconda3 to the PATH to use `conda` command.
+    export PATH="$TMP_DIR/miniconda3/bin:$PATH"
+
+    # Initialize conda. This replaces calling `conda init bash`.
+    # Conda init command requires a shell restart which should not be done on BK.
+    source "$TMP_DIR/miniconda3/etc/profile.d/conda.sh"
+}
+
+run_sanity_check() {
+    python_version=$1
```
Comment (aslonnie) [inline .buildkite/release-automation/verify-macos-wheels.sh:36]: `PYTHON_VERSION`, and maybe add `local` (optional), and double quote `"$1"`

File: .buildkite/release-automation/verify-macos-wheels.sh [hunk 5]
```diff
@@ -0,0 +1,64 @@
+#!/bin/bash
+
+set -euo pipefail
+
+set -x
+
+python_versions=("3.8" "3.9" "3.10" "3.11")
+
+# Check arguments
+if [ "$#" -ne 1 ]; then
```
Comment (aslonnie) [inline .buildkite/release-automation/verify-macos-wheels.sh:10]: nit: use double bracket

also, you do not have to double quote `$#` double quotes are for strings. numbers do not need double quoting (though double quoting is also fine)

`if [[ $# -ne  1 ]]; then`

more details:

https://google.github.io/styleguide/shellguide.html#quoting
https://google.github.io/styleguide/shellguide.html#test----and---

probably just read the full styling guide :)

**Review by aslonnie — APPROVED** [review]
mostly look good. just some small things.

let me know when this is ready to merge.

File: .buildkite/release-automation/verify-macos-wheels.sh (hunk 2 above)
Comment (khluu) [reply .buildkite/release-automation/verify-macos-wheels.sh]: fixed!

File: .buildkite/release-automation/verify-macos-wheels.sh (hunk 3 above)
Comment (khluu) [reply .buildkite/release-automation/verify-macos-wheels.sh]: fixed!

File: .buildkite/release-automation/verify-macos-wheels.sh (hunk 5 above)
Comment (khluu) [reply .buildkite/release-automation/verify-macos-wheels.sh]: fixed!

**Review by aslonnie — APPROVED** [review]

**Pull request merged**